### PR TITLE
Fixing event feed filtering

### PR DESCRIPTION
**Pull request by lancewf**
Changing event feed filters to filter on fields and not tags. The event feed service was performing filters with 'event type', 'requestor', and 'organizations' on the tags field. This does not work because if any of the above filters have similar values, the filtered events would have been incorrectly returned. For example, if an event had a 'requestor' = 'profile' and 'event type' = 'scan job' then when filtering for 'event type' = 'profile' this event would be returned incorrectly.

The Change required a migration of the eventfeed-1-feeds Elasticsearch index. Because of the migration, other mapping changes were made to prevent future migrations. 

Pull out the `FormatFilters` function from three components (applications, event-feed, and config-mgmt) into the lib directory.

### CODE REVIEW

File: components/applications-service/pkg/server/server.go [hunk 1]
```diff
@@ -210,7 +211,7 @@ func (app *ApplicationsServer) GetServices(
 	ctx context.Context, request *applications.ServicesReq,
 ) (*applications.ServicesRes, error) {
 
-	filters, err := params.FormatFilters(request.GetFilter())
+	filters, err := stringutils.FormatFilters(request.GetFilter())
```
Comment (lancewf) [inline components/applications-service/pkg/server/server.go]: Pulling out the `FormatFilters`.

File: components/config-mgmt-service/params/parameters.go [hunk 2]
```diff
@@ -93,7 +89,7 @@ func ConvertParamToActionBackend(parameter string) string {
 // 	"roles": ["lalala"],
 // ]
 func FormatNodeFilters(filters []string) (map[string][]string, error) {
-	return formatFilters(filters, ConvertParamToNodeRunBackend)
+	return stringutils.FormatFiltersWithKeyConverter(filters, ConvertParamToNodeRunBackend)
```
Comment (lancewf) [inline components/config-mgmt-service/params/parameters.go]: Pulling out the `FormatFilters`.

File: components/config-mgmt-service/params/parameters_test.go [hunk 3]
```diff
@@ -25,33 +24,6 @@ func TestFormatFiltersEmptyArray(t *testing.T) {
...
-			filters, err := subject.FormatNodeFilters(test.filters)
-
-			assert.NotNil(t, err)
-			assert.Equal(t, errors.InvalidParameter, err.(*errors.StandardError).Type)
-			assert.Contains(t, err.(*errors.StandardError).Message, "Invalid filter")
-			assert.Contains(t, err.(*errors.StandardError).Message, "format: key:value")
-			assert.Nil(t, filters)
-		})
-	}
-}
```
Comment (lancewf) [inline components/config-mgmt-service/params/parameters_test.go]: Pulling out the `FormatFilters`.

File: components/event-feed-service/pkg/integration_test/feed_test.go [hunk 4]
```diff
@@ -222,128 +222,6 @@ func TestEventFeedReturnOnlyEventsWithinDateRange(t *testing.T) {
 	}
 }
 
-func TestEventFeedFilterTags(t *testing.T) {
```
Comment (lancewf) [inline components/event-feed-service/pkg/integration_test/feed_test.go]: The tags field is not going to be used for filtering, so there is currently no way to add `org_name` to an event.

File: components/event-feed-service/pkg/migration/migration.go [hunk 5]
```diff
@@ -10,6 +10,7 @@ import (
 
 var (
 	eventFeedZeroIndexName    = "eventfeed-0-feeds"
+	eventFeedOneIndexName     = "eventfeed-1-feeds"
```
Comment (lancewf) [inline components/event-feed-service/pkg/migration/migration.go]: Adding the migration to the new 'eventfeed-2-feeds' index

File: components/event-feed-service/pkg/persistence/feed-elastic.go [hunk 6]
```diff
@@ -134,124 +135,88 @@ func (efs ElasticFeedStore) CreateFeedEntry(entry *util.FeedEntry) (bool, error)
...
 }
 
+// GetFeed - get event feed entries with the provided query constraints
 func (efs ElasticFeedStore) GetFeed(query *util.FeedQuery) ([]*util.FeedEntry, int64, error) {
 	exists, err := efs.indexExists(IndexNameFeeds)
 	if err != nil {
 		logrus.Warnf("Could not get feed; error: %+v", err)
 		return nil, 0, err
 	}
+	if !exists {
```
Comment (lancewf) [inline components/event-feed-service/pkg/persistence/feed-elastic.go]: Made this more go like 😄

File: components/event-feed-service/pkg/persistence/feed-elastic.go [hunk 7]
```diff
@@ -267,28 +232,29 @@ func (efs ElasticFeedStore) getCounts(q *util.FeedSummaryQuery) (map[string]int6
...
-				// if there's only one value, construct an AND query (one MUST with multiple values for "tags"), e.g.:
-				// "must":[
-				//    {"term":{"tags":"org_1"},
-				//	  {"term":{"tags":"org_2"}
-				//  ]
-				value := values[0]
-				boolQuery.Must(olivere.NewTermQuery(field, value))
-			}
+
+	for field, values := range filters {
```
Comment (lancewf) [inline components/event-feed-service/pkg/persistence/feed-elastic.go]: Change the filtering to filter off a specific field for each type of filter and not only the tags field.

File: components/event-feed-service/pkg/persistence/mappings.go [hunk 8]
```diff
@@ -27,72 +27,112 @@ type Mapping struct {
...
-				"type":"text"
+			"type":"text"
 		},
 		"object_name":{
-				"type":"text"
+			"type":"text"
 		},
 		"object_object_type":{
-				"type":"keyword"
+			"fields": {
```
Comment (lancewf) [inline components/event-feed-service/pkg/persistence/mappings.go]: Added these mappings to allow for wildcard and suggestions.

File: components/event-feed-service/pkg/server/feed.go [hunk 9]
```diff
@@ -59,17 +64,23 @@ func (f *Feeds) GetFeed(req *event_feed.FeedRequest) ([]*util.FeedEntry, int64,
...
 }
 
-func (f *Feeds) GetFeedSummary(countCategory string, filters []string, start time.Time, end time.Time) (*util.FeedSummary, error) {
+func (f *Feeds) GetFeedSummary(countCategory string, filters []string,
+	start time.Time, end time.Time) (*util.FeedSummary, error) {
 
 	// remove count category from the filters array
 	filters = util.Remove(filters, countCategory)
 
+	mapFilters, err := util.FormatFilters(filters)
```
Comment (lancewf) [inline components/event-feed-service/pkg/server/feed.go]: Convert the URL parameters to Elasticsearch filters.

File: components/event-feed-service/pkg/util/util.go [hunk 10]
```diff
@@ -145,46 +159,11 @@ func BytesToString(data []byte) string {
...
 // map[string][]string [
 // 	"environment": ["adios","hola"],
-//  "cookbook": ["awesome"],
-//  "roles": ["lalala"]
+// 	"cookbook": ["awesome"],
+// 	"roles": ["lalala"],
 // ]
-//
 func FormatFilters(filters []string) (map[string][]string, error) {
-	filterMap := make(map[string][]string, len(filters))
```
Comment (lancewf) [inline components/event-feed-service/pkg/util/util.go]: Pulling out the FormatFilters.

File: lib/stringutils/filter_parser.go [hunk 11]
```diff
@@ -1,45 +1,49 @@
...
+// 	"roles": ["lalala"],
 // ]
-// TODO: (afiune) Migrate this to a common go package that other teams can consume
 func FormatFilters(filters []string) (map[string][]string, error) {
+	return FormatFiltersWithKeyConverter(filters, func(k string) string { return k })
+}
+
+// FormatFiltersWithKeyConverter The same as FormatFilters with the function keyNameConverter that
+// converts the keys from aliases.
+func FormatFiltersWithKeyConverter(filters []string,
```
Comment (lancewf) [inline lib/stringutils/filter_parser.go]: Pulling out the FormatFilters.

File: lib/stringutils/filter_parser.go [hunk 12]
```diff
@@ -1,45 +1,49 @@
...
 // map[string][]string [
-// 	"status": ["critical"],
-// 	"service_name": ["redis","postgres"],
-// 	"application": ["cafe"],
-// 	"environment": ["prod"]
+// 	"environment": ["adios","hola"],
+// 	"cookbook": ["awesome"],
+// 	"roles": ["lalala"],
 // ]
-// TODO: (afiune) Migrate this to a common go package that other teams can consume
```
Comment (afiune) [inline lib/stringutils/filter_parser.go]: I see what you did there! Super nice Lance 💯

tenor-245281441

File: components/event-feed-service/pkg/persistence/feed-elastic.go (hunk 6 above)
Comment (phiggins) [reply components/event-feed-service/pkg/persistence/feed-elastic.go]: If this is something the service can't recover from, failing fast seems like a good idea 👍

**Pull request merged**